integration/image/delta_test.go

```diff
@@ -45,6 +47,75 @@ func TestDelta(t *testing.T) {
 	}
 }
 
+// TestDeltaCreateDestinationLock triggers a delta generation job, waits for it
+// to start and removes the destination image.
+//
+// Until 1e83de47e32ac0caf3b4e02094aeb76da28b90b3 this would remove unprocessed
+// destination layers and break delta generation.
+func TestDeltaCreateDestinationLock(t *testing.T) {
+	var (
+		base   = "debian:8"
+		target = "debian:10"
+		delta  = "debian:delta"
+	)
+
+	var (
+		err    error
+		rc     io.ReadCloser
+		ctx    = context.Background()
+		client = testEnv.APIClient()
+	)
+
+	pullBaseAndTargetImages(t, client, base, target)
+
+	rc, err = client.ImageDelta(ctx,
+		base,
+		target,
+		types.ImageDeltaOptions{
+			Tag: delta,
+		})
+	if err != nil {
+		t.Fatalf("Creating delta: %s", err)
+	}
+	defer rc.Close()
+
+	var (
+		waitFingerprinting = make(chan struct{}, 100)
+		waitDelta          = make(chan struct{})
+	)
+	go func() {
+		defer close(waitFingerprinting)
+		defer close(waitDelta)
+		for br := bufio.NewReader(rc); ; {
+			line, err := br.ReadString('\n')
+			if err != nil {
+				if err == io.EOF {
+					break
+				}
+				t.Fatal(err)
+			}
+			if strings.Contains(line, "Fingerprinting") {
+				waitFingerprinting <- struct{}{}
+			}
+			if strings.Contains(line, "layer does not exist") {
+				t.Fail()
+			}
+		}
+	}()
+
+	<-waitFingerprinting
+	_, err = client.ImageRemove(ctx, target, types.ImageRemoveOptions{})
```

Comment:
If I understand correctly, that ignored return value contains a slice with the images that were actually removed. In this case, it would make sense to one more assert to check that this list is empty, wouldn't it?

Reply:
based on your suggestion, I think I improved the tests overall by adding a few assertions to make sure what we're looking at actually conforms to the delta we want to create :)